Extract refactoring

MFlisar opened the pull request:
So far this seems to work, I think I did not forget anything.

* `StaticSetup` singleton is awful, I know, but it was the fastest solution without changing too much
* small bugfix - search view close behaviour changed => close it if no text is in search view, clear text only otherwise

TODO because of refactoring:
* resource strings are just copied now, will be merged, but still, should be cleaned...

MFlisar commented:
I would appreciate that someone looks over it before I continue fixing/improving other things, otherwise the refactoring changes and the improvements/fixes will get mixed up too much...

gsantner commented:
* Please don't put anything UI related in Application object. Same to logic: Shouldn't be inside App. The App object should be as slim as possible.
* No need for the `I` prefix for Interfaces. E.g. IAppItem is still a AppItem.
* Dependencies: Didn't look into it, but be sure to **not** include libs that depend on gplay services. It should still be buildable for F-Droid.

I don't have to for a full review, but these are my first thoughts. As this is a major change you will need to wait for a merge for approval from either BennyKok or dkanada too. In a rewrite it's better to fix errors in design from the past instead of doing the same thing again - getting as fast as possible in.

MFlisar commented:
* I'll change the thing with the I prefix
* Dependencies: nothing changed there anyways
* App - I don't see any problems with the current solution, but it's no problem to change that. Will move to code to the `Home` class...

Generally, as discussed with BennyKok, this is just my first refactoring to make the app's core reuseable as I want to use the launcher functionality in my app only with my custom data layer, items and so on. Then when testing and improving the usage in my app I will make further adjustments/improvements, the first step is just to offer the interfaces and have some reusable core without changing too much. Correcting design errors can be done afterwards and are not the goal of this pull request, although I may do this as well and remove the static instances in the views for example and a few things I already saw that are not working.

gsantner commented:
quick response: `App` - summed up: bad practise to put data in it. I dont have some trustful source in my list but you can find a lot on it. Not saying Home is/was ideal too, features and class design came over time.

Yeah, it's not about correcting old errors, but it shouldn't introduce new ones (if possible)

MFlisar commented:
Improvement: get rid of fast adapter ViewHolder factory by updating dependencies ;-)

I'll use this in my own app which is in production already, improvements will be done as well, you can be sure about that.

gsantner commented:
👍

dkanada commented on app/src/main/java/com/benny/openlauncher/widget/SearchBar.java:
```diff
@@ -111,7 +115,7 @@ public void onClick(View v) {
         searchInput.setVisibility(View.GONE);
         searchInput.setBackground(null);
-        searchInput.setHint("Search...");
+        searchInput.setHint(R.string.search_hint);
```
Looks like we missed a string here, thanks.

MFlisar replied:
I think I've seen even a few more... Btw, as I think you made the preferences  ever seen this library: https://github.com/dkunzler/esperandro ?

I personally use it for years already and it offer following:
* you only define preferences in one file (no string definitions anymore)
* the annotation processor does the rest, including creating string resources
* if desired, the generated preference manager class does caching, this means, getting values again and again will be fast without the need of having to store them in a local variable

No real overhead for the app as everything is done at compile time.

Additionally I have one question:
Is the mixed usage of local AppSettings objects and AppSettings.get() in the views desired or did it just happen to be so?

gsantner replied:
Currently the AppSettings objects are just local ones for easier access (`appSettings` vs.  `AppSettings.get()`)

MFlisar replied:
So it should be save to always use the local one in the views? When I'm looking them through I can clean that.

gsantner replied:
Yeah, as long as initalization is ensured

dkanada replied:
Now would be a good time to remove all local instances if we want to use the static version every time, I think that would work.

MFlisar replied:
I'll do that

dkanada commented on app/src/main/java/com/benny/openlauncher/widget/AppItemView.java:
```diff
@@ -20,17 +20,17 @@
  */
 
-public class AppItemView extends View implements Drawable.Callback {
+public class AppItemView extends View implements Drawable.Callback, com.benny.openlauncher.core.interfaces.AppItemView {
```
I think Drawable.Callback is an interface as well, are we moving all interfaces into a separate folder now?

MFlisar replied:
I want to have the ability to allow custom items, that's why the item is defined as interface for the core. I want to change it even further to have some abstract function like `displayIcon(ImageView iv)` or so, so that the implementation of the item can use Glide, Picasso or similar as well and the item won't hold a big drawable in memory.

dkanada replied:
ah okay

gsantner replied:
Using glide sounds good, wouldn't mind if it would be used in core too. maybe via the displayIcon abstraction, but still in there

dkanada commented on core/src/main/java/com/benny/openlauncher/core/interfaces/DialogHandler.java:
```diff
@@ -0,0 +1,25 @@
+    void showDeletePackageDialog(Context context, DragEvent dragEvent);
+
+    interface IOnAddAppDrawerItem
+    {
```
Brackets are on the same line as the declaration in this project.

dkanada commented on core/src/main/java/com/benny/openlauncher/core/widget/Dock.java:
```diff
@@ -49,14 +49,14 @@ public void init() {
+        int columns = StaticSetup.get().getAppSettings().getDockSize();
         setGridSize(columns, 1);
-        List<Item> dockItems = Home.db.getDock();
+        List<Item> dockItems = home.db.getDock();
```
I think the desktop initializes in a different location than the dock, it might not pass in Home as a variable. We should modify one or the other so they both have a similar location in the code.

BennyKok commented:
is anyone here ok with the naming of interface with prefix "I"

dkanada commented:
I would just add Interface at the end, similar to MinibarEditActivity or HomeActivity.

BennyKok commented:
oh you guys want e.g. AppInterface or just simply App?

dkanada commented:
I think AppInterface would be best.

gsantner commented:
Those are 2 differnt kind of things. Class - Interface <--> Activity - View.

Just take a look at how it's done in java [lists](https://stackoverflow.com/questions/2814805/java-interfaces-implementation-naming-convention/2814926#2814926)

* List - interface for the "conceptual" object
* ArrayList - concrete implementation of interface
* LinkedList - concrete implementation of interface
* AbstractList - abstract "partial" implementation to assist custom implementations

BennyKok commented:
so AbstractApp?

gsantner commented:
May be okay, but that doesn't mean everything needs to be prefixed with **Abstract**.   
At the case of Application object: Neither. App should be very slim, don't implement interfaces in there.

BennyKok commented:
I noticed that the new branch has been closed, so did you guys mean to merge this to the master branch when this is done?

BennyKok commented:
Oh actually I was outdated, the latest commit in MFlisar's clone of this repo has already remove the I prefix from the new interfaces …

gsantner commented on core/src/main/java/com/benny/openlauncher/core/interfaces/AppDeleteListener.java:
```diff
@@ -0,0 +1,9 @@
+ */
+
+public interface AppDeleteListener<T extends App> {
+    void onAppDeleted(T app);
```
What is the purpose of AppDeleted? App stopped?  
Does AppDeleted mean app was uninstalled?

dkanada replied:
uninstalled I think

BennyKok replied:
Yes, the app manager will find the missing app, (I don't remember exactly) but
since you brought up this question, a long time, I though why don't I include a simple callback so I can easily know when an app is deleted, however,  it turns out that this is not required

BennyKok replied:
I am not sure if there are any usage of this interface now

gsantner replied:
Maybe generalizing the Update event would suit better, and give it an attribute like 
- Update
- Removed from something (folder, desktop, hidden, ..)
- Uninstall
- Installed

BennyKok replied:
Wow that's a good point.

MFlisar replied:
it's used in Open Launcher already and is called if an app is uninstalled to update the drawer and similar

MFlisar replied:
Just saw that there have been some answers before mine. The `AppManager` takes care to notify anyone who has a list of apps to update the list: https://github.com/OpenLauncherTeam/openlauncher/blob/master/app/src/main/java/com/benny/openlauncher/util/AppManager.java#L232

It simply does compare the last known apps with the newly loaded apps

gsantner commented:
I think I got somewhat confused out of the whole `App` stuff. If App is related to e.g. a installed item/view thats okay of course to do whatever. My comments where about the Application Object (the one that `exends android.app.Application`).

dkanada commented:
So is <ClassName>Interface okay then?

gsantner commented:
No; E.g. SomethingListener<Class>

BennyKok commented:
Guys I think we broke the record, this is by far the longest conversation in this repo.

MFlisar commented:
So that you all know what I do next:

* make the core independent of any fast adapter item => I will use my own items everywhere that use glide to load images memory efficient. `AppManager` loads all apps into memory (which is fine) and then loads all app icons, which means that those will stay in memory. Recycling views won't help much here, if the model holds the `Drawable`.
* remove the local `AppSetting` instances from the core's views
* adjust the `StaticSetup` class to have understandable function names, current names were just directly derived without thinking about it

MFlisar commented:
**DONE**
* moved some hardcoded strings to resources
* removed all local references to the static global settings
* defined ids for fast adapter items => fast adapter internally relys on this, naver use different layouts without different ids
* StaticSetup renamed to Setup and functions cleaned a little bit
* Copied the WidgetHost view from my app => long press on widgets should work now, even if those are invalid (after restoring the app or if the widget's app was uninstalled) + no need to move to fire the long press event
* Moving icon on another one to create a folder => Crash in GroupIconDrawable + Icons were not visible in folder => solved by getting the icons from the AppManager.App instead of the app's files directory where no image is found...

gsantner commented on app/src/main/java/com/benny/openlauncher/viewutil/GroupIconDrawable.java:
```diff
@@ -107,10 +108,14 @@ public void draw(Canvas canvas) {
-        canvas.drawBitmap(icons[1], null, new RectF(iconSizeDiv2 + padding, padding, iconSize - padding, iconSizeDiv2 - padding), paint2);
-        canvas.drawBitmap(icons[2], null, new RectF(padding, iconSizeDiv2 + padding, iconSizeDiv2 - padding, iconSize - padding), paint2);
-        canvas.drawBitmap(icons[3], null, new RectF(iconSizeDiv2 + padding, iconSizeDiv2 + padding, iconSize - padding, iconSize - padding), paint2);
+        if (icons[0] != null)
```
if () {
statement
}

gsantner commented on app/src/main/java/com/benny/openlauncher/widget/SearchBar.java:
```diff
@@ -13,200 +11,32 @@
 
 /**
- * Created by BennyKok on 4/27/2017.
+ * Created by Michael on 25.06.2017.
```
I think we can drop such lines (any name, other files too) and apply apache 2.0 license notice instead

dkanada replied:
I don't even think we need the license notice considering there's one in the root directory, makes things easier.

gsantner commented on app/src/main/java/com/benny/openlauncher/widget/SearchBar.java:
```diff
@@ -13,200 +11,32 @@
-
-        void onCollapse();
+    protected void startApp(Context context, App app)
+    {
```
braces

gsantner commented on app/src/main/res/values-de/strings.xml:
```diff
@@ -19,7 +19,7 @@
     <string name="wallpaper_unable_to_blur">Kann Blureffekt nicht anwenden.</string>
 
-    <string name="group">Ohne Namen : - (</string>
+    <string name="group">Ordner</string>
```
If it is in fact a folder, lets use `<string name="folder">Folder</string>` in default translation (and `folder">Ordner`)

gsantner commented on core/build.gradle:
```diff
@@ -0,0 +1,43 @@
+    buildToolsVersion "25.0.3"
+
+    defaultConfig {
+        minSdkVersion 16
```
Can the core be lowered in minSdkVersion? (Even if OpenLauncher app maybe cannot support it, others might)

gsantner commented on core/src/main/java/com/benny/openlauncher/core/interfaces/App.java:
```diff
@@ -0,0 +1,14 @@
+import android.graphics.drawable.Drawable;
+
+/**
+ * Created by Michael on 25.06.2017.
```
lic

gsantner commented on core/src/main/java/com/benny/openlauncher/core/interfaces/DialogHandler.java:
```diff
@@ -0,0 +1,23 @@
+    void showEditDialog(Context context, T item, IOnEditDialog resultHandler);
+    void showDeletePackageDialog(Context context, DragEvent dragEvent);
+
+    interface IOnAddAppDrawerItem {
```
**I**On --> `OnAppDrawerItemAddedListener` maybe?

dkanada replied:
Yeah the listener suffix could work.

gsantner commented on core/src/main/java/com/benny/openlauncher/core/manager/DefaultSettings.java:
```diff
@@ -0,0 +1,196 @@
+ * Created by flisar on 27.06.2017.
+ */
+
+public class DefaultSettings implements SettingsManager {
```
nice idea

gsantner commented on core/src/main/java/com/benny/openlauncher/core/manager/Setup.java:
```diff
@@ -0,0 +1,113 @@
+    // Class and singleton
+    // ----------------
+
+    private static Setup mSetup = null;
```
Setup setup

dkanada replied:
^

gsantner commented on core/src/main/java/com/benny/openlauncher/core/manager/Setup.java:
```diff
@@ -0,0 +1,113 @@
+
+    private static Setup mSetup = null;
+
+    public static boolean wasInitialised()
```
braces

gsantner commented on core/src/main/java/com/benny/openlauncher/core/manager/Setup.java:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    public static Setup get() {
+        if (mSetup == null)
```
missing braces

gsantner commented on core/src/main/java/com/benny/openlauncher/core/viewutil/CircleDrawable.java:
```diff
@@ -0,0 +1,101 @@
+            canvas.drawBitmap(hidingOldIcon ? iconToFade : icon, iconSize / 2 - iconSizeReal / 2, iconSize / 2 - iconSizeReal / 2, paint2);
+            canvas.restore();
+
+            if (currentScale == 0)
```
braces

gsantner commented on core/src/main/java/com/benny/openlauncher/core/viewutil/CircleDrawable.java:
```diff
@@ -0,0 +1,101 @@
+
+        if (iconToFade != null) {
+            canvas.save();
+            if (hidingOldIcon)
```
braces or ternary

gsantner commented on core/src/main/java/com/benny/openlauncher/core/viewutil/CircleDrawable.java:
```diff
@@ -0,0 +1,101 @@
+            if (currentScale == 0)
+                hidingOldIcon = false;
+
+            if (!hidingOldIcon && scaleStep == 1)
```
2

gsantner commented on core/src/main/java/com/benny/openlauncher/core/widget/AppDrawerVertical.java:
```diff
@@ -34,14 +33,14 @@
-    private static int itemWidth;
-    private static int itemHeightPadding;
+    public static int itemWidth;
+    public static int itemHeightPadding;
```
anything against `private int`? is static as performance reason?

MFlisar replied:
yes - variables are used in `DrawerItem.ViewHolder`. I would anyway suggest removing any global static variable from such classes, does not look clean imho. Should be passed through to whoever needs them if possible...

gsantner commented on core/src/main/java/com/benny/openlauncher/core/widget/BaseSearchBar.java:
```diff
@@ -0,0 +1,233 @@
+ */
+
+public class BaseSearchBar extends FrameLayout {
+    public TextView searchClock;
```
private?

gsantner commented on core/src/main/java/com/benny/openlauncher/core/widget/GroupPopupView.java:
```diff
@@ -83,56 +81,51 @@ public boolean showWindowV(final Item item, final View itemView, final DesktopCa
         for (int x2 = 0; x2 < cellSize[0]; x2++) {
             for (int y2 = 0; y2 < cellSize[1]; y2++) {
-                if (y2 * cellSize[0] + x2 > item.items.size() - 1) {
+                if (y2 * cellSize[0] + x2 > item.getGroupItems().size() - 1) {
```
as above in strings: Is it group or it's specific child: folder?
@BennyKok @dkanada

dkanada replied:
Not sure what you mean but this looks correct, I think the calculation is for the size of the container and card view which need the number of items in the group.

gsantner replied:
Its not about the line of code. I asked if there is a difference between a folder and a group. If its only use is folder (e.g. on desktop) than maybe call it folder instead

gsantner commented on core/src/main/java/com/benny/openlauncher/core/widget/WidgetView.java:
```diff
@@ -0,0 +1,140 @@
+    private final int THRESHOLD;
+    private OnLongClickListener longClick;
+
+    private boolean mHasPerformedLongPress;
```
hasPerformedLongPress

gsantner commented on core/src/main/java/com/benny/openlauncher/core/widget/WidgetView.java:
```diff
@@ -0,0 +1,140 @@
+    @Override
+    public boolean onInterceptTouchEvent(MotionEvent ev) {
+
+        if (mOnTouchListener != null && mOnTouchListener.onTouch(this, ev))
```
braces

gsantner commented on core/src/main/java/com/benny/openlauncher/core/widget/WidgetView.java:
```diff
@@ -0,0 +1,140 @@
+                float diffY = Math.abs(mLongPressDownY - ev.getY());
+                //L.d("onInterceptTouchEvent: diffX = %f | diffY = %f | THRESHOLD = %d", diffX, diffY, THRESHOLD);
+                if (diffX >= THRESHOLD || diffY >= THRESHOLD)
+                {
```
braces

gsantner commented:
Maybe apply Refactor -> Reformat in Android Studio, might fix many of the points

dkanada commented on app/src/main/java/com/benny/openlauncher/model/AppItem.java:
```diff
@@ -0,0 +1,82 @@
+
+public class AppItem extends AbstractItem<AppItem, AppItem.ViewHolder> implements com.benny.openlauncher.core.interfaces.AppItem<AppItem, AppItem.ViewHolder>
+{
+    public AppManager.App app;
```
also need to move open bracket to the same line as the declaration

dkanada commented on app/src/main/java/com/benny/openlauncher/model/Item.java:
```diff
@@ -182,23 +178,86 @@ public void writeToParcel(Parcel out, int flags) {
     }
 
+    @Override
+    public int getId() {
```
pretty sure this method is overriding the method from the interface, not sure though

gsantner replied:
sorry, member variable behind, not the method

MFlisar commented:
@gsantner this happens because I use a different brace format in my own projects and have defined this in android studio, because I don't like having the open brace on the same line as the statement... will correct that though

MFlisar commented:
Everything I've not commented is done now. I removed all "Created by ..." statements as well.

* `minSdk` may be possible to be lowered, not sure though
* As the `Setup` has functions to create items, glide can be used in those items or not. Or even picasso or anything else. It's open to the implementation currently already

gsantner commented on core/src/main/java/com/benny/openlauncher/core/widget/BaseSearchBar.java:
```diff
@@ -0,0 +1,229 @@
+    }
+
+    protected void startApp(Context context, App app)
+    {
```
braces

BennyKok commented:
@MFlisar big thanks too

MFlisar commented:
you're welcome

dkanada commented:
This looks good to me, let's tag one more release for bug fixes and then we can merge the changes.

gsantner commented:
@dkanada @BennyKok 
any doubts?

dkanada commented:
No issues here, just pushed a new release so we can merge now.

gsantner commented:
@dkanada The merge is targeted against a seperate branch, not master. Do you want it to merge to master immediatly, or try it out a little first?

MFlisar commented:
Before using the master, wait for my next pull request

I'm cleaning the resources and have fixed a few small things and corrected one bug I've introduced
